File: mqtt_esp32_simulator/sensor_smoke.py
import random
from sensor import Sensor

class SmokeSensor(Sensor):
    def __init__(self, sensor_id, name):
        super().__init__(sensor_id, name, "smoke")
        self.smoke_detected = False
        self.obscuration = round(random.uniform(0.0, 0.02), 2)

    def generate_data(self):
        self.obscuration = self.generate_random_value(self.obscuration, 0.0, 10.0, 0.05)
        self.smoke_detected = True if self.obscuration > 0.3 else False

        return {
            "smoke_detected": self.smoke_detected,
            "obscuration": round(self.obscuration, 2),
        }

File: mqtt_flutter_app/lib/sensors/sensor_smoke.dart
import 'package:flutter/material.dart';
import 'sensor.dart';

class SmokeSensor extends Sensor {
  bool? smokeDetected;
  double? obscuration;

  SmokeSensor({
    required super.id,
    required super.name,
    super.lastUpdated,
    this.smokeDetected,
    this.obscuration,
  }) : super(type: SensorType.smoke);

  @override
  Widget buildWidget(BuildContext context) {
    return Row(
      children: [
        Expanded(
          flex: 1,
          child: Icon(Icons.local_fire_department,),
        ),
        Expanded(
          flex: 6,
          child: Text(name),
        ),
        Expanded(
          flex: 3,
          child: Text('$obscuration'),
        ),
      ],
    );
  }
}
